git_machete/tests/functional/test_machete.py

```diff
@@ -26,6 +30,182 @@ def get_head_commit_hash() -> str:
+        for pull in self.pulls:
+            if pull['number'] == pull_no:
+                return pull
+        return None
+
+
+class MockGithubAPIResponse:
+    def __init__(self, status_code: int, response_data: Union[List[Dict[str, Any]], Dict[str, Any]]) -> None:
+        self.response_data: Union[List[Dict[str, Any]], Dict[str, Any]] = response_data
+        self.status_code: int = status_code
+
+    def read(self) -> bytes:
+        return json.dumps(self.response_data).encode()
+
+
+class MockGithubAPIRequest:
+    def __init__(self, github_api_state: MockGithubAPIState) -> None:
+        self.github_api_state: MockGithubAPIState = github_api_state
+
+    def __call__(self, url: str, headers: Dict[str, str] = None, data: Union[str, bytes, None] = None, method: str = '') -> "MockGithubAPIResponse":
+        self.parsed_url: ParseResult = urlparse(url, allow_fragments=True)
+        self.parsed_query: Dict[str, List[str]] = parse_qs(self.parsed_url.query)
+        self.json_data: Union[str, bytes] = data
+        self.return_data: Optional[Union[List[Dict[str, Any]], Dict[str, Any]]] = None
+        self.headers: Dict[str, str] = headers
+        return self.handle_method(method)
+
+    def handle_method(self, method: str) -> "MockGithubAPIResponse":
+        if method == "GET":
+            return self.handle_get()
+        elif method == "PATCH":
+            return self.handle_patch()
+        elif method == "POST":
+            return self.handle_post()
+        else:
+            return self.make_response_object(HTTPStatus.METHOD_NOT_ALLOWED, [])
+
+    def handle_get(self) -> "MockGithubAPIResponse":
+        if self.parsed_url.path.endswith('pulls'):
+            full_head_name: Optional[List[str]] = self.parsed_query.get('head')
+            if full_head_name:
+                head: str = full_head_name[0].split(':')[1]
+                for pr in self.github_api_state.pulls:
+                    if pr['head']['ref'] == head:
+                        return self.make_response_object(HTTPStatus.OK, [pr])
+                return self.make_response_object(HTTPStatus.NOT_FOUND, [])
+            else:
+                return self.make_response_object(HTTPStatus.OK, self.github_api_state.pulls)
+        elif self.parsed_url.path.endswith('user'):
+            return self.make_response_object(HTTPStatus.OK, [self.github_api_state.user])
+        else:
+            return self.make_response_object(HTTPStatus.NOT_FOUND, [])
+
+    def handle_patch(self) -> "MockGithubAPIResponse":
+        if 'issues' in self.parsed_url.path:
+            return self.update_issue()
+        elif 'pulls' in self.parsed_url.path:
+            return self.update_pull_request()
+        else:
+            return self.make_response_object(HTTPStatus.NOT_FOUND, [])
+
+    def handle_post(self) -> "MockGithubAPIResponse":
+        assert not self.parsed_query
+        if self.parsed_url.path.endswith('issues'):
+            return self.update_issue()
```

Comment:
One thing, can't we just call `create_issue/pull_request` in `handle_post`? this way `if not pull_no/issue_no` could probably be removed from `update_...` methods

Reply:
Without further changes, we can't. This is because POST is used to add reviewers or assignees to pulls/issues while sending only a fraction of data which is necessary to create pull/issue, which is considered by put  `MockGihubAPIRequest` as an update.

Reply:
Ok ☹️ let's leave as is then